Fate reservations moved out of memory

kevinrr888 opened the pull request:
closes #4131
Fate reservations were moved out of memory. This is a WIP PR* for one part that is needed for having Fate be distributed in the future.

- Reservations for MetaFateStore were moved out of memory into ZooKeeper
- Reservations for UserFateStore were moved out of memory into the Accumulo Fate table
	- Is an additional column which just indicates whether the FateId of that row is reserved or not
- Added test MultipleStoresIT
	- Tests the new functionality of how reservations are stored
	- I also included a couple tests in MultipleStoresIT that are testing the reserve/unreserve functionality and that it still works as intended. This can be moved from this class if desired.
- Verified existing tests in the fate test package (FateIT, FateOpsCommandsIT, FateStoreIT, ZooMutatorIT, FateMutatorImplIT, UserFateStoreIT) still pass

*There are still a few things I have marked as TODO that I would like input/suggestions on. These are marked in the code as "TODO 4131".

kevinrr888 commented:
Marking this as ready for review, since it may not be reviewed otherwise

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -150,6 +150,39 @@ protected void create(FateId fateId, FateKey fateKey) {
+  @Override
+  public Optional<FateTxStore<T>> tryReserve(FateId fateId) {
+    // TODO 4131 should this throw an exception if the id doesn't exist (status = UNKNOWN)?
+    FateMutator.Status status = newMutator(fateId).putReservedTx(fateId).tryMutate();
```
This code needs to be refactored to use a UUID similar to the zookeeper code.  The UUID covers the  problem of a write making it the server, but the client not knowing the write made it.  Both zookeeper and accumulo can have this problem and both can have a similar solution.  The following examples show how the UUID can help address this problem.

 1. FATE1 exists and is not currently reserved.
 2. Thread1 attempts reserve FATE1 by writing `isreserved` only if current value is `notreserved`.
 3. Thread2 attempts reserve FATE1 by writing `isreserved` only if current value is `notreserved`.
 4. One of the writes succeeds and the other fails.  However before the tablet server can respond to Thread1 or Thread2, it dies.  This causes both thread to receive an unknown status.
 5. After getting the UNKNOWN status, Thread1 reads FATE1 and sees it marked reserved however it does not know if it reserved FATE1 or another thread.
 6. After getting the UNKNOWN status, Thread2 reads FATE1 and sees it marked reserved however it does not know if it reserved FATE1 or another thread.

In the above example, FATE1 is marked reserved but no thread will be working on it.  If instead of setting a value of `isreserved` we instead set a per thread UUID, then the above situation could look like the following.

 1. FATE1 exists and is not currently reserved. Its reserved column has a value of empty string.
 2. Thread1 wants to reserve FATE1 and generates UUID1 for this purpose
 3. Thread2 wants to reserve FATE1 and generates UUID2 for this purpose
 4. Thread1 attempts reserve FATE1 by writing UUID1 as a value for the reserved column only if current value is empty.
 5. Thread2 attempts reserve FATE1 by writing UUID2 as a value for the reserved column only if current value is empty.
  6. The write for Thread2 succeeds and the write for Thread1 fails because the col val is not empty.  However before the tablet server can respond to Thread1 or Thread2, it dies.  This causes both thread to receive an unknown status.
  7.  After getting the UNKNOWN status, Thread1 reads FATE1 reserved column and sees UUID2 so it knows it did not reserve it.
  8. After getting the UNKNOWN status, Thread2 reads FATE1 reserved column and sees UUID2 so it knows it did reserve it.

In addition to the UUID mentioned above, the persisted reservation will need another field and that is a process lock id. The UUID mentioned above is scoped very narrowly, it scoped to an individual reservation attempt and its used to handle the case of server applying write and then dying.  The process lock id would be used to detect reservation that are held by a process that is no longer alive.  The following is an example of how all of this works together.

 1. A manager process starts in Accumulo and acquires a lock in ZK, let calls it PL1
 2. The manager process reserves FATE1 setting the reservation column to UUID1:PL1.  UUID1 is scoped to this reservation thread/code block as mentioned above.
 3. After successfully reserving (could have seen unknown status and check for UUID1) the manager process dies
 4. A new manager process is started and acquires a lock in ZK, let calls it PL2
 5. The new manager process starts FATE, which starts a periodic task to look for reservations held by dead process.  The periodic task sees FATE1 is reserved by PL1 which no longer has lock in ZK and is there presumed dead.
 6. The periodic task issues a condtional mutation (or conditional update for ZK) to delete the reserved value for FATE1 only if the currently value is UUID1:PL1

Taking all of this together, may want to do the following.

 1. Create a new type that represent the reservation value, maybe like 
     ```java
           class FateReseveration {
                  UUID reservationId;
                  ZooUtil.LockID processID; // TODO not sure if this is the best type for this
           }
     ```
  2. Use the above type internall in the User and Meta Fatestores for their reservations
  3. Add methods to FateStore like `Stream<FateReservation> getActiveReservations()` and `void deleteReservation(FateReservation)`
  4. Add a periodic task to Fate that calls the two methods above and looks active reservations held by dead processes and then deletes them.
  5. When creating a new FateStore instance, pass in the lock of the process where it is running.  For now this would mean when the manager creates a new FateStore instance it will pass in its lock. This will be used to set the processID on reservations.

keith-turner replied:
The manager lock is [here](https://github.com/apache/accumulo/blob/0edc48f910f25e5273d62b1d91a7993bf9529b8e/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L217) and from that can access [ServiceLock.getLockID()](https://github.com/apache/accumulo/blob/0edc48f910f25e5273d62b1d91a7993bf9529b8e/core/src/main/java/org/apache/accumulo/core/lock/ServiceLock.java#L594) and [ServiceLock.isLockHeld()](https://github.com/apache/accumulo/blob/0edc48f910f25e5273d62b1d91a7993bf9529b8e/core/src/main/java/org/apache/accumulo/core/lock/ServiceLock.java#L654).  So can persist the LockId w/ reservations, although not sure of the best way to do that.  Then the periodic test that looks for reservations held by dead processes it can call ServiceLock.isLockHeld.

kevinrr888 replied:
Thank you for this detailed explanation. This was really helpful for https://github.com/apache/accumulo/pull/4524/commits/e6671fbb903da53014bd64e53ad125351add6a04

kevinrr888 commented:
I have made the following changes in https://github.com/apache/accumulo/pull/4524/commits/e6671fbb903da53014bd64e53ad125351add6a04:
- Combined the LockID and reservation attempt UUID into one new object: FateReservation
- Now cover edge case with tryReserve() where a write may make it to the server, but the server dies before a response is received
- New FATE Thread: DeadReservationCleaner which deletes reservations held by Managers that have since died
- Created new classes: ColumnValueMappingIterator and ReservationMappingIterator. ColumnValueMappingIterator abstracts out the common functionality of ReservationMappingIterator and StatusMappingIterator. ReservationMappingIterator is an iterator used for determining if the reservation column for a FateId has a FateReservation set or not
- Expanded/improved/simplified MultipleStoresIT tests

There are also some comments/questions labeled "TODO 4131" that I would like input on.

I also have to resolve the merge conflicts which I will do now.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/MultipleStoresIT.java:
```diff
@@ -0,0 +1,571 @@
+          .withConstructor(ClientContext.class, String.class, ZooUtil.LockID.class)
+          .withArgs(client, tableName, lock1).addMockedMethod("isDeadReservation").createMock();
+    } else {
+      mockedStore1 = EasyMock.createMockBuilder(MetaFateStore.class)
+          .withConstructor(String.class, ZooReaderWriter.class, ZooCache.class,
+              ZooUtil.LockID.class)
+          .withArgs(FATE_DIR, zk, client.getZooCache(), lock1).addMockedMethod("isDeadReservation")
+          .createMock();
```
Not sure about this, but the following pattern may simplify this test over mocking where liveLocks is a `Set<ZooUtil.LockID>` that the test can add and remove things to as needed.

```suggestion
      mockedStore1 = new MetaFateStore<>(FATE_DIR, zk, client.getZooCache(), lock1) {
        @Override
        public boolean isDeadReservation(FateReservation reservation) {
          return !liveLocks.contains(reservation.getLockID()); 
        }
      };
```

Another thing to consider is passing in a `Predicate<ZooUtil.LockID>` to the Fate store constructors that can be used to check if a lock is live. Then the `isDeadReservation()` method could be dropped an stores could use this predicate internally.  Test could pass in a predicate that checks a set.  The real code could pass in a predicate taht calls the ServiceLock code.

kevinrr888 replied:
The predicate is a great idea and cleans up the code quite a bit; added

keith-turner reviewed (commented):
Still looking at this PR. Posting the comments I have so far.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -106,13 +112,102 @@ public FateId create() {
+    } catch (Exception e) {
+      // Exception thrown, so node doesn't exist, so it is not reserved
+      isReserved = false;
+    }
```
Looking at getNode(), the NNE is handled in that method, so I simplified isReserved()

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -224,9 +323,14 @@ public void setStatus(TStatus status) {
-            NodeExistsPolicy.OVERWRITE);
-      } catch (KeeperException | InterruptedException e) {
+        zk.mutateExisting(getTXPath(fateId), currSerializedData -> {
+          NodeValue currNodeVal = new NodeValue(currSerializedData);
```
Fixed

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -315,6 +420,30 @@ public List<ReadOnlyRepo<T>> getStack() {
+              // Remove the FateReservation from the NodeValue to unreserve
+              return new NodeValue(currNodeVal.status, null, currFateKey).serialize();
+            } else {
+              // possible this is running a 2nd time in zk server fault conditions and its first
```
Added

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -502,4 +520,14 @@ protected Serializable deserializeTxInfo(TxInfo txInfo, byte[] data) {
   }
+
+  /**
+   * TODO 4131 this is a temporary method used to create a dummy lock when using a FateStore outside
```
Yes, that sounds like a great change. And yes you are correct about some of the fate admin utils requiring the Manager to be down, so that would be helpful for those as well.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -151,6 +154,76 @@ protected void create(FateId fateId, FateKey fateKey) {
+
+  @Override
+  public boolean isReserved(FateId fateId) {
+    return newMutator(fateId).requireReserved().tryMutate().equals(FateMutator.Status.ACCEPTED);
```
Understood, fixed

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -272,8 +340,12 @@ public FateInstanceType type() {
+      super(fateId);
+    }
+
+    private FateTxStoreImpl(FateId fateId, FateReservation reservation) {
```
That sounds like a good addition

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -151,6 +154,76 @@ protected void create(FateId fateId, FateKey fateKey) {
+      status = newMutator(fateId).requireReserved(reservation).tryMutate();
+      if (status.equals(FateMutator.Status.ACCEPTED)) {
+        return Optional.of(new FateTxStoreImpl(fateId, reservation));
+      }
```
Yeah, I wasn't sure about this mutation. This is a much better way to do this; fixed.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/Fate.java:
```diff
@@ -313,6 +317,26 @@ private void undo(FateId fateId, Repo<T> op) {
+    @Override
+    public void run() {
+      while (keepRunning.get()) {
+        store.deleteDeadReservations();
```
After the DeadReservationCleaner dies, the Manager continues to run normally. I changed it to be a critical thread. When looking at this, I noticed that none of the other Fate threads are critical. It seems that maybe they should be? I can create a follow on issue if so. If the WorkFinder or any of the actual workers (TransactionRunner) die, Manager will continue to run normally.

keith-turner replied:
That would be good to open a follow on issue about making all of the fate threads critical.

kevinrr888 commented on test/src/main/java/org/apache/accumulo/test/fate/MultipleStoresIT.java:
```diff
@@ -0,0 +1,571 @@
+      createFateTable(client, tableName);
+      mockedStore1 = EasyMock.createMockBuilder(UserFateStore.class)
+          .withConstructor(ClientContext.class, String.class, ZooUtil.LockID.class)
+          .withArgs(client, tableName, lock1).addMockedMethod("isDeadReservation").createMock();
```
For now, I went with your next suggestion with passing the predicate to the stores. This is not as end-to-end because the lock is not actually in ZK but may be good enough for now. Can be done in follow on issue if desired.

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -151,6 +157,95 @@ protected void create(FateId fateId, FateKey fateKey) {
+        scanner.setRange(getRow(fateId));
+        scanner.fetchColumn(TxColumnFamily.RESERVATION_COLUMN.getColumnFamily(),
+            TxColumnFamily.RESERVATION_COLUMN.getColumnQualifier());
+        FateReservation persistedRes = scanner.stream()
+            .filter(entry -> FateReservation.isFateReservation(entry.getValue().toString()))
+            .map(entry -> FateReservation.from(entry.getValue().toString())).findFirst()
+            .orElse(null);
+        if (persistedRes != null && persistedRes.equals(reservation)) {
+          return Optional.of(new FateTxStoreImpl(fateId, reservation));
+        }
```
```suggestion
        return scanner.stream().map(entry -> entry.getValue().toString())
            .filter(FateReservation::isFateReservation).map(FateReservation::from)
            .filter(persistedRes -> persistedRes.equals(reservation)).findFirst()
            .map(persistedRes -> new FateTxStoreImpl(fateId, reservation));
```

These could be combined

kevinrr888 replied:
I wrote what would be the equivalent stream needed, and it was pretty complicated. I personally think the way it is now is a bit more readable/easy to follow. Since it would be the same, I'm gonna leave it as is.

DomGarguilo reviewed (commented):
Just some small changes. Feel free to take or leave any

keith-turner reviewed (commented):
Still review this, posting the comments I have made so far.  Hoping to circle back to this really soon like later today or tomorrow and finish reviewing.  

I went through and resolved all the comments I made last time, all of those updates look good.  The one I left unresolved were only because those made need follow on issues opened so I did not want to forget about those.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -151,6 +157,95 @@ protected void create(FateId fateId, FateKey fateKey) {
+        scanner.fetchColumn(TxColumnFamily.RESERVATION_COLUMN.getColumnFamily(),
+            TxColumnFamily.RESERVATION_COLUMN.getColumnQualifier());
+        FateReservation persistedRes = scanner.stream()
+            .filter(entry -> FateReservation.isFateReservation(entry.getValue().toString()))
+            .map(entry -> FateReservation.from(entry.getValue().toString())).findFirst()
```
Looking into this a bit more, I think the reservation column may always exist and have a not reserved value.  So is the filter needed for that case?

kevinrr888 replied:
Yes, the reservation column always exists and the filter is needed for the case where it is not reserved. The value stored if the fate id is not reserved is FateMutatorImpl.NOT_RESERVED = "", and this is the value set when a new fate id is created and when a reserved fate id is unreserved. Thinking about this again, it would probably be better to just delete/create the column as needed instead of storing a not reserved value. What do you think?

kevinrr888 replied:
Yeah, I also think deleting would be better. I will make this change.

keith-turner replied:
>If going with the NOT_RESERVED value approach it would be good to make the validation of the column value more strict, will make a comment about that elsewhere.

Nevermind the code does not work exactly the way I thought it did. I was thinking when reading reservations could do something like the following

```java
if(value look like <lock id>:<uuid) {
  // process reserved
} else if (value is NOT_RESERVED) {
  //process not reserved
} else {
   throw new IllegalArgumentException("invalid reseved column value");
}
```

However the is not a place in the code that reads reservations in this way.  Was thinking the code should fail if the reserved column value was something like '`1234` because it does not look like a reservation and it also does not look like NOT_RESERVED

keith-turner replied:
If the reservation parsing code is not really strict and the code writing reservations has a bug, then could end with the case were reservation is silently ignored because it did not have the correct format.  Would be better to fail rather than ignore things that are not recognized.  Some of the code seems to have the following pattern

```java
f(value look like <lock id>:<uuid) {
  // process reserved
} else {
  // assume not reserved, but could have been an incorrectly written reservation
}
```

kevinrr888 replied:
Ah I see. Yeah I don't think I ever check if something is reserved like that and instead use a conditional mutation in FateMutatorImpl to reserve/unreserve. Should be pretty straightforward to change to create/delete the column on reserve/unreserve

kevinrr888 replied:
You're referring to MetaFateStore where I have that pattern correct? I don't think UserFateStore would have this issue but I could be wrong

kevinrr888 replied:
I'm not sure I follow how the reservation could be incorrectly written. I could just be missing/misunderstanding something but for MetaFateStore where I do this check, it checks if a FateReservation object is present and that can only take on the form <lock id>:<uuid>. Creating the FateReservation object handles these checks of making sure only a valid FateReservation of the form <lock id>:<uuid> is created.

keith-turner replied:
Made this comment as an example

https://github.com/apache/accumulo/pull/4524/files?diff=unified&w=1#r1672972761

I am not sure if MetaFateStore has any problems, was only thinking through the UserFateStore code.  Ideally both should be very strict w/ reservation parsing.

kevinrr888 replied:
Ah I understand now. Yes that should (but currently doesn't) throw an error if the entry is not NOT_RESERVED and not a valid FateReservation str. Since I will be changing the column to only ever include a valid FateReservation (column created and deleted as needed), I will just change this to throw an error if it sees anything but a FateReservation string. Thanks for the clarification

keith-turner replied:
I had made a comment elsewhere about avoiding using `toString()` for serialization and suggested using `getSerialized()` instead  When I made that comment I thought the impl of `toString()` and `getSerialized()` were the same.  Now I realize they are different.  Maybe part of the solution to this is converging UserFateStore and MetaFateStore on the same serialization impl and making the de-serialization as strict as possible. Would that work?  If NOT_RESERVED is dropped seems like that makes it easier for them to have same ser/deser impl.

kevinrr888 replied:
Yeah I think that would work and sounds like a good idea. It would clean up/get rid of some code. Don't think converting the FateReservation to/from a string would be needed anymore.

kevinrr888 replied:
I attempted to refactor UserFateStore to create/delete the RESERVATION column on reserve/unreserve, however, I ran into an issue that I could not figure out, so I ended up reverting this change. Multiple threads would occassionally reserve the same fate id. I could not figure out how/why this was occuring. I think the switch to deleting the column should be made, but maybe as a follow on or after discussion about the potential cause of this. Maybe I could put up a new commit with just this change to easily see the difference and what the issue might be.

keith-turner replied:
> Multiple threads would occassionally reserve the same fate id.

Not sure what the problem was, but its kinda neat you detected that bug.  Where you running an IT when you saw multiple threads reserve a transaction?

kevinrr888 replied:
Yes, I was running my new `MultipleStoresIT.testDeadReservationsCleanup()` test, which would occasionally fail with my new changes. The test works as follows:

1) We have a fate object fate1 which is used to have 4 (the default MANAGER_FATE_THREADPOOL_SIZE) workers reserve 4 transactions with lock1
2) We then simulate that the imaginary manager dies and a new one starts, so lock1 is no longer held instead lock2 is now held
3) A new fate2 is created which spins up 4 more workers which should eventually reserve the 4 transactions so long as the DeadReservationCleaner thread unreserves the transactions reserved with lock1

Took a while to track down, but found that when waiting for the transactions to be reserved with lock2, multiple threads would sometimes reserve the same transaction, so not all 4 transactions would be reserved causing the test to fail. Found that the issue was due to the change from storing an unreserved value in the RESERVATION column to just deleting/creating the column on unreserve/reserve.

I could not figure out why or how this could occur with these changes. I could potentially put a new commit up at some point with these changes to see if you could find anything wrong with how I am doing it.

keith-turner replied:
This conversion is really long, but it seems like everything in it is resolved.  @kevinrr888 can you resolve this if you also don't see anything outstanding in the conversations?

kevinrr888 replied:
Yes, resolved. This conversation will be linked in the new issue to address this change as context.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/user/FateMutatorImpl.java:
```diff
@@ -79,6 +82,33 @@ public FateMutator<T> putCreateTime(long ctime) {
+    Condition condition = new Condition(TxColumnFamily.RESERVATION_COLUMN.getColumnFamily(),
+        TxColumnFamily.RESERVATION_COLUMN.getColumnQualifier()).setValue(NOT_RESERVED);
+    mutation.addCondition(condition);
+    TxColumnFamily.RESERVATION_COLUMN.put(mutation, new Value(reservation.toString()));
```
Its nice to avoid calling toString() when its return value is used with serialized data as the expectations around toString() are not well defined. For example someone could change the toString impl to improve a log message or something like that and not realize the implications for serialized data.

```suggestion
    TxColumnFamily.RESERVATION_COLUMN.put(mutation, new Value(reservation.getSerialized()));
```

kevinrr888 replied:
Yes, good call. I now only use the serialized value

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/user/FateMutatorImpl.java:
```diff
@@ -79,6 +82,33 @@ public FateMutator<T> putCreateTime(long ctime) {
+  @Override
+  public FateMutator<T> putUnreserveTx(FateStore.FateReservation reservation) {
+    Condition condition = new Condition(TxColumnFamily.RESERVATION_COLUMN.getColumnFamily(),
+        TxColumnFamily.RESERVATION_COLUMN.getColumnQualifier()).setValue(reservation.toString());
```
```suggestion
        TxColumnFamily.RESERVATION_COLUMN.getColumnQualifier()).setValue(reservation.getSerialized());
```

keith-turner commented on core/src/main/java/org/apache/accumulo/core/logging/FateLogger.java:
```diff
@@ -175,6 +176,21 @@ public Optional<FateTxStore<T>> createAndReserve(FateKey fateKey) {
+
+      @Override
+      public void deleteDeadReservations() {
+        store.deleteDeadReservations();
```
Probably would not go here, but it would be nice to have consistent logging somehow for each dead reservation that is detected and deleted.

keith-turner replied:
Not sure if this is workable,  maybe the method could take a consumer that is passes found dead reservations to  like : `deleteDeadReservations(Consumer<FateReservation> foundDead)` then this could add logging for those.  Alternatively could just put the same log stmts in each impl of the store.

kevinrr888 replied:
Yes, nice suggestion. For simplicity in this new commit, I just added the same log message for both impls.

keith-turner reviewed (commented):
I have reviewed everything but the test changes, plan on looking at those tomorrow and then I will be finished reviewing.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/Fate.java:
```diff
@@ -73,9 +75,12 @@ public class Fate<T> {
+  private final ExecutorService deadResCleanerExecutor;
 
   private static final EnumSet<TStatus> FINISHED_STATES = EnumSet.of(FAILED, SUCCESSFUL, UNKNOWN);
+  private static boolean userDeadReservationCleanerRunning = false;
+  private static boolean metaDeadReservationCleanerRunning = false;
```
> Should these be Atomic? Or is it fine to keep them as-is?

Right now, there is only one Manager which runs, so these do not need to be Atomic, but the goal is to have multiple managers in the future, so not sure if this should be changed now or in the future.

> Can these static variables could be removed? Do these help testing in someway? In the manager we can assume that only a single instance of each type is created. If we wanted to guard against multiple instances per type, then a more general check could be done that is not limited to dead reservation cleaners.

The purpose of these was to avoid creating more than 1 DeadReservationCleaner running per store type. As I currently have this, the first Fate object to be created per store type will be running a DeadReservationCleaner (in the future, this probably should be changed as running on the first Fate object created may not be the best approach). In the Manager, only one Fate object will be created per store type, so this isn't a necessary check now, but isn't the plan in the future to have multiple Fate instance running per store type, so in that case we'd only ever want 1 DeadReservationCleaner running? I was premtively handling this case. What do you have in mind regarding a more general check? Or should I not be handling this case yet?

kevinrr888 replied:
Left as-is until further clarification

keith-turner replied:
In #3964  I was thinking of having this [PartitionData](https://github.com/apache/accumulo/blob/208e39927310f8ff7dd5978f2794e26bee6be1b4/core/src/main/java/org/apache/accumulo/core/manager/PartitionData.java) object that would be passed to Fate from the manager.  Fate could use that to partition fate operations to run and to determine if it should run a singleton service using the `PartitionData.shouldRun` method.   Was thinking that PatitionData could be passed to all the managers high level services (like Fate, CompactionCoordinator, TabletGroupWatcher) and that all could use it to partition their data in their own way.

With something like that, in test if you create 3 fate instances then could pass them `ParitionData(0,3)`, `ParitionData(1,3)`, and `ParitionData(2,3)`.   It probably does not makes sense to add that in this PR as it will not partition the fate ids at this point.

For this PR we could simplify things and have a boolean argument on the Fate constructor that indicates if the dead reservation detection should be started and not have any static variable that attempt to enforce one per process.

kevinrr888 replied:
:+1: re the PartitionData object
Got it, I've changed it to take a boolean argument for whether the cleaner should run.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -151,6 +157,95 @@ protected void create(FateId fateId, FateKey fateKey) {
+      scanner.fetchColumn(TxColumnFamily.RESERVATION_COLUMN.getColumnFamily(),
+          TxColumnFamily.RESERVATION_COLUMN.getColumnQualifier());
+      scanner.stream()
+          .filter(entry -> FateReservation.isFateReservation(entry.getValue().toString()))
```
If we had buggy code that was writing out a reservation that looked like `<lockid>:<uuid>:<uuid>` I think `isFateReservation` would return false for this (not completley sure, hard to follow the curly braces in the diff).  I think it will split the string on `:` and return false if does not have two fields.  Would be better to make this more strict and throw an exception when unexpected data like this is seen.

kevinrr888 replied:
Addressed. Now, isFateReservation() throws an error if it's passed anything unnexpected (only expects a NOT_RESERVED value or a valid FateReservation serialization).

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/Fate.java:
```diff
@@ -360,7 +380,27 @@ public Fate(T environment, FateStore<T> store, Function<Repo<T>,String> toLogStr
+      } else if (store.type() == FateInstanceType.META) {
+        metaDeadReservationCleanerRunning = true;
+      }
+    }
```
Added

kevinrr888 replied:
Resolved as no longer applicable: no longer check the store type

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/Fate.java:
```diff
@@ -360,7 +380,27 @@ public Fate(T environment, FateStore<T> store, Function<Repo<T>,String> toLogStr
+        metaDeadReservationCleanerRunning = true;
+      }
+    }
+    this.deadResCleanerExecutor = deadResCleanerExecutor;
```
yes, it won't ever be used. I did it this way for a simpler shutdown method (no need to check for null), but you are right it would be better if this wasn't created in the first place. Changed.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/FateStore.java:
```diff
@@ -107,11 +118,155 @@ interface FateTxStore<T> extends ReadOnlyFateTxStore<T> {
+          DataOutputStream dos = new DataOutputStream(baos)) {
+        dos.writeUTF(lockID.serialize("/"));
+        dos.writeUTF(reservationUUID.toString());
+        dos.close();
```
Looking at existing code, this was done elsewhere (see FateKey or CompactionConfigStorage for example), and is what I based my code off of since I was doing the same thing. I also didn't understand why the close() was done early, but I left it in to be safe.

keith-turner replied:
I suspect close is called here to ensure all data is flushed in the DataOutputStream before calling `baos.toByteArray()` which is under the DataOutputStream.  Seems like calling `flush()` would be ok.

kevinrr888 replied:
I'll leave it as-is unless there is a significant difference between the two. If it's changed here, it should probably be changed everywhere this is done, which shouldn't be done in this PR.

keith-turner replied:
@DomGarguilo if you think it would be best to call flush everywhere can you please create and issue and resolve this comment?  There are so many comments on this issue, we are trying to get them all resolved, if no issue is needed can just resolve it.

DomGarguilo replied:
Since the code functions correctly and its not a big deal. I'll just resolve this without any further changes needing to be made

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -106,13 +112,95 @@ public FateId create() {
+        String txUUIDStr = strTxId.split("_")[1];
+        FateId fateId = FateId.from(fateInstanceType, txUUIDStr);
+        if (isReserved(fateId)) {
+          FateReservation reservation = getNode(fateId).reservation.orElseThrow();
```
Addressed

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -168,18 +173,16 @@ public void runnable(AtomicBoolean keepWaiting, Consumer<FateId> idConsumer) {
                 }
-                return !reserved.contains(fateId);
               }
+              return !isReserved(fateId);
```
Addressed

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -298,7 +299,13 @@ private Optional<FateId> create(FateKey fateKey) {
   @Override
   public Optional<FateTxStore<T>> createAndReserve(FateKey fateKey) {
-    FateId fateId = fateIdGenerator.fromTypeAndKey(getInstanceType(), fateKey);
+    // TODO 4131 not confident about this new implementation of createAndReserve.
```
Implemented in User/MetaFateStore.createAndReserve(). Thank you for this detailed description.

kevinrr888 replied:
As it is now, UFS.createAndReserve is pretty lengthy and both implementations could probably share common code. Was mostly interested if the impls are good. Can do code cleanup after.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -114,38 +131,26 @@ public static Object deserialize(byte[] ser) {
-          throw new IllegalStateException(e);
-        }
+    Preconditions.checkState(!_getStatus(fateId).equals(TStatus.UNKNOWN),
+        "Attempted to reserve a tx that does not exist: " + fateId);
```
Good catch, fixed.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -361,90 +356,108 @@ public Optional<FateTxStore<T>> createAndReserve(FateKey fateKey) {
+      verifyReserved(false);
 
-        long countBefore = unreservedNonNewCount.getCount();
+      int currNumCallers = concurrentStatusChangeCallers.incrementAndGet();
```
Good suggestion, fixed.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/FateStore.java:
```diff
@@ -107,11 +118,155 @@ interface FateTxStore<T> extends ReadOnlyFateTxStore<T> {
+        String[] fields = fateReservationStr.split(":");
+        if (fields.length == 2) {
+          return LOCKID_PATTERN.matcher(fields[0]).matches()
+              && UUID_PATTERN.matcher(fields[1]).matches();
```
No longer applicablevvvvvvvvv

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -106,13 +112,95 @@ public FateId create() {
+  public void deleteDeadReservations() {
+    for (Map.Entry<FateId,FateReservation> entry : getActiveReservations().entrySet()) {
+      FateId fateId = entry.getKey();
+      FateReservation reservation = entry.getValue();
```
Nice improvement. Added

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -224,9 +316,21 @@ public void setStatus(TStatus status) {
+          NodeValue currNodeVal = new NodeValue(currSerializedData);
+          // Ensure the FateId is reserved in ZK, and it is reserved with the expected reservation
+          if (currNodeVal.isReserved()
+              && currNodeVal.reservation.orElseThrow().equals(this.reservation)) {
```
Yeah I do that check quite a bit... I think it's a good change and is more clear. Added

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -55,7 +60,7 @@
 //TODO use zoocache? - ACCUMULO-1297
 //TODO handle zookeeper being down gracefully - ACCUMULO-1297
-
+// TODO 4131 noticed this class is not in the fate.zookeeper package. Should it be?
```
Yeah that's what I thought. This can be done in a follow on issue/PR. This can be kept unresolved until then

kevinrr888 commented:
@keith-turner, @DomGarguilo 
I have addressed all of the review changes, and this PR is ready for re-review. However, some tests are no longer passing (e.g., ComprehensiveIT). I believe this is an issue with my new implementation of createAndReserve. I believe the issue is with the UserFateStore implementation, but could be both. Posting this new commit now to get the other changes out, and to potentially receive feedback on what might be wrong with createAndReserve(). I will continue to look into what the issue might be in the meantime. Please take a look at my new review comments and my new commit message:

- Stricter check of the column value for fate reservations. If anything unexpected is seen, an error is now thrown.
- Simplified MetaFateStore.getActiveReservations() to only read from ZooKeeper once.
- Combined the two scans that were done in AbstractFateStore.runnable() into one. This meant adding FateReservation to FateIdStatus and refactoring Meta/UserFateStore.getTransactions().
- No longer use/store a string representation of the FateReservation (was used in UserFateStore). Now, only the serialized value is used. This keeps the usage of FateReservation consistent across Meta and UserFateStore. It was also unneccessary, so simplifies code.
- Moved AbstractFateStore.createAndReserve() implementation into Meta and UserFateStore, and rewrote the impl for each to work with the new way reservations are stored. This also allowed me to delete methods that were only used by AFS.createAndReserve(): create(FateId, FateKey), getStatusAndKey(FateId), create(FateKey).
- Fixed how concurrentStatusChangeCallers was decremented in AbstractFateStore.waitForStatusChange()
- Small change to MetaFateStore.deleteDeadReservations() to avoid reading from ZK unnecessarily
- Added isReservedBy() method to MetaFateStore.NodeValue to avoid code duplication and make the code more clear.
- Since the FateIdStatus now has the FateReservation, realized Meta and UserFateStore.getActiveReservations() could now be simplified to just call list(). This also made the impls the same, so moved to AbstractFateStore. This also made me realize that I had put getActiveReservations() method signature in FateStore, but would be better suited for ReadOnlyFateStore, so moved it there.
- Deleted FateStore.isReserved(FateId)... No longer needed/used
- Moved UNKNOWN status check in AbstractFateStore.reserve() into waiting loop
- Now log when a dead reservation is detected and deleted
- Minor change to Fate: no longer create the executor for the dead reservation cleaner if it's not going to be used

kevinrr888 commented:
I have found and addressed the bug in my most recent commit. This PR is now ready for full review.
The tests verified to still pass: the sunny day tests, all fate tests (ZooMutatorIT, FateMutatorImplIT, UserFateStoreIT, FateInterleavingIT, FateIT, FateOpsCommandsIT, ComprehensiveFlakyFateIT, DeleteRowsFlakyFateIT, ManagerRepoIT, and MultipleStoresIT (new test added)) except FateStoreIT. FateStoreIT reflects a method that has been removed in this PR, and want to receive feedback about the removal of this method before changes are made to this test. Have a "TODO 4131" marking this so this is not forgotten.

keith-turner reviewed (commented):
Reviewed everything but the test.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -67,27 +76,36 @@ public FateId fromTypeAndKey(FateInstanceType instanceType, FateKey fateKey) {
-    this.reserved = new HashSet<>();
-    this.deferred = new HashMap<>();
+    this.deferred = Collections.synchronizedMap(new HashMap<>());
+    this.lockID = Objects.requireNonNullElseGet(lockID, AbstractFateStore::createDummyLockID);
```
This is simple to address in this PR; changed to only accept non-null and changed to caller to create the dummy lock instead

keith-turner replied:
ok sounds good.  I am working though comments and resolving them.  Left some comments open that seemed like they needed a follow on issue.   Once issues are opened can resolve those comments.  We may need an issue about making the fate utils get a lock.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -104,19 +109,108 @@ public FateId create() {
+    FateReservation reservation = FateReservation.from(lockID, UUID.randomUUID());
+
+    try {
+      byte[] newSerNodeVal = zk.mutateExisting(getTXPath(fateId), currSerNodeVal -> {
```
Ah okay, I thought we would want to throw the exception in the case the node doesn't exist. Changed and added check to MultipleStoresIT.testReserveUnreserve(). I see now that the tryReserve() for UFS and MFS were different for this case: UFS would return empty optional, MFS would throw an error, so this change is good and keeps them consistent. I remember noticing this difference a while ago when I was writing this, but must have forgot to write down/ask which functionality is expected. Glad this was spotted again.

keith-turner replied:
I looked into adding a test for this in elasticity branch, but found it was not worthwhile if the behavior will change. Could add at test to `FateStoreIT.testAbsent()` in this PR that does the following.

```java
    var optTxStore = store.tryReserve(fateId);
    assertTrue(optTxStore.isEmpty());
```

kevinrr888 replied:
Yeah, I considered adding the check to that test, but chose to add it to MultipleStoresIT.testReserveUnreserve() since there it also makes sure it doesn't affect the current active reservations at all.
Are you suggesting I also have the check in testAbsent()? Or that I move the check?

keith-turner replied:
The test in MultipleStoresIT is fine, I had not looked at that yet.  I was looking in the elasticity branch to see if this case  was tested there.  As long tryReserve  for non existent is tested somewhere its good.

kevinrr888 replied:
Gotcha. I added that check to testAbsent() in this PR. Doesn't hurt to have it tested multiple places and makes to have in testAbsent()

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -104,19 +109,108 @@ public FateId create() {
+            return null;
+          }
+        });
+      } catch (KeeperException | InterruptedException | AcceptableThriftTableOperationException e) {
```
Good catch. Fixed

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -224,9 +322,20 @@ public void setStatus(TStatus status) {
+            NodeValue newNodeValue = new NodeValue(status, currFateReservation, currFateKey);
+            return newNodeValue.serialize();
+          } else {
+            throw new IllegalStateException("Either the FateId is not reserved in ZK, or it is"
```
Yes, I must have forgotten it here. Added and included the reservation in the log as well. Double checked AFS, UFS, and MFS that everywhere something is logged or thrown regarding a fate id that the fate id is included in the message, so no follow on issue needed.

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -363,13 +498,23 @@ protected Stream<FateIdStatus> getTransactions(Set<TStatus> statuses) {
+            Suppliers.memoize(() -> {
+              NodeValue zkNode = getNode(fateId);
+              return new Pair<>(zkNode.status, zkNode.reservation);
+            });
```
Simplified

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/MetaFateStore.java:
```diff
@@ -421,14 +569,30 @@ private Optional<FateKey> deserializeFateKey(DataInputBuffer buffer) throws IOEx
+    private Optional<FateReservation> deserializeFateReservation(DataInputBuffer buffer)
+        throws IOException {
+      int length = buffer.readInt();
+      if (length > 0) {
```
Added check to both deserializeFateKey() and deserializeFateReservation()

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/user/FateStatusFilter.java:
```diff
@@ -54,8 +55,15 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
+    try {
+      var tstatus = ReadOnlyFateStore.TStatus.valueOf(v.toString());
+      return valuesToAccept.contains(tstatus);
+    } catch (IllegalArgumentException e) {
```
Reworked iterator. Thanks for the info, this is much nicer than what I had

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -120,35 +127,161 @@ public FateId getFateId() {
+          Optional<FateKey> fateKeySeen = Optional.empty();
+          Optional<FateReservation> reservationSeen = Optional.empty();
+
+          for (Entry<Key,Value> entry : scanner.stream().collect(Collectors.toList())) {
```
haha fixed

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/fate/user/UserFateStore.java:
```diff
@@ -120,35 +127,161 @@ public FateId getFateId() {
+            verifyFateKey(fateId, fateKeySeen, fateKey);
+            // NEW/unseeded transaction and not reserved, so we can allow it to be reserved
+            // we tryReserve() since another thread may have reserved it since the scan
+            txStore = tryReserve(fateId);
```
Added

kevinrr888 commented:
@keith-turner - Addressed all review comments, this PR is ready for re-review. 

See new commit bd1f17472c8d289878e19d7121f9fc9347e95596

I also fixed FateStoreIT to work with the new impl of createAndReserve(). A method was deleted in 6760035608387ae537af7050106c9645aac19aae that was no longer used which FateStoreIT relied on. Upon fixing this test, noticed that the original impl of createAndReserve expected an error to be thrown in a certain situation, but in the new implementation, an empty optional was returned instead. Fixed the new impl to match the previous impl for this case. This meant changing UFS.createAndReserve and MFS.createAndReserve. 

Also verified existing Fate tests still pass: ZooMutatorIT,FateMutatorImplIT,UserFateStoreIT,FateInterleavingIT,FateIT,FateOpsCommandsIT,ComprehensiveFlakyFateIT,DeleteRowsFlakyFateIT,ManagerRepoIT,MultipleStoresIT,FateStoreIT
and verified sunny day tests still pass.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/user/RowFateStatusFilter.java:
```diff
@@ -53,19 +56,28 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
+  protected boolean filter(Text currentRow, List<Key> keys, List<Value> values) {
+    for (int i = 0; i < keys.size(); i++) {
+      Key key = keys.get(i);
+      if (key.getColumnQualifier()
+          .equals(FateSchema.TxColumnFamily.STATUS_COLUMN.getColumnQualifier())
```
```suggestion
      if (FateSchema.TxColumnFamily.STATUS_COLUMN.hasColumns(key)
```

kevinrr888 replied:
Changed

keith-turner commented on core/src/main/java/org/apache/accumulo/core/fate/AbstractFateStore.java:
```diff
@@ -57,8 +57,8 @@
 // TODO 4131 should probably add support to AbstractFateStore, MetaFateStore,
-// and UserFateStore to accept null lockID and zooCache (maybe make these fields
-// Optional<>). This could replace the current createDummyLockID(). This support
+// and UserFateStore to accept null lockID (maybe make this field Optional<>).
```
When I have a PR w/ a lot of TODOs sometimes I will try to open an issue for each TODO and remove it before merging.  Could do that for this PR but do not have to, if not doing it  then it would be good to use a more unique TODO which I think is being done here?  Do all of the TODOs have the 4131 suffix?  If so then it will be easy to find them with grep.  Then we just need an open  issue that records the fact that all `TODO 4131` need to be done.

kevinrr888 replied:
Yes, I have all the TODOs marked `TODO 4131`. I was planning on creating new issues and deleting all the `TODO 4131` when this PR is ready to be merged.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/metrics/fate/meta/MetaFateMetrics.java:
```diff
@@ -59,7 +60,8 @@ public void registerMetrics(MeterRegistry registry) {
     try {
-      return new MetaFateStore<>(getFateRootPath(context), context.getZooReaderWriter());
+      return new MetaFateStore<>(getFateRootPath(context), context.getZooReaderWriter(),
+          AbstractFateStore.createDummyLockID(), null);
```
This store is being created for metrics, so it may be using it in a readonly way and may never reserve. If it never reserves then it does not need a lock, not even a dummy lock.  Not something for this PR, but for this case I think we want to somehow create a store obj where attempts to reserve will be non functional and throw an exception.

kevinrr888 replied:
Yeah, can leave this unresolved for making a follow on issue. I put the dummy lock in many places where a store was created, without regard to whether reservations would actually occur. A follow on issue can address this.

keith-turner commented:
I am finally looking through the test changes right now.   Everything else looks good.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/MultipleStoresIT.java:
```diff
@@ -0,0 +1,553 @@
+      store = new MetaFateStore<>(FATE_DIR, zk, lock, null);
+    }
+
+    assertThrows(IllegalStateException.class, () -> store.reserve(fakeFateId));
```
Could verify  error messages contains the fate id.

```suggestion
    var err = assertThrows(IllegalStateException.class, () -> store.reserve(fakeFateId));
    assertTrue(err.getMessage().contains(fakeFateId.canonical()));
```

kevinrr888 replied:
Added

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/MultipleStoresIT.java:
```diff
@@ -0,0 +1,553 @@
+    testDeadReservationsCleanup(FateInstanceType.USER);
+  }
+
+  private void testDeadReservationsCleanup(FateInstanceType storeType) throws Exception {
```
This test looks good.  Looking over this test realized we need another test in a follow on PR. Need a test that checks that when the reservation does not exists in the persisted store that every update fails.  Test would look something like the following.

```
    FateStore.FateTxStore<?> txStore = store1.reserve(fateId);
    // TODO delete the reservation in persisted store
    assertThrows(Exception.class, ()->txStore.push(repo));
    assertThrows(Exception.class, ()->txStore.setStatus(ReadOnlyFateStore.TStatus.IN_PROGRESS));
    assertThrows(Exception.class, ()->txStore.setTransactionInfo(t,v));
    //etc test all methods that write
```

For now may only be able to make this work with the UserFateStore as the current zookeeper schema does not support the conditional checks of reservations on updates.

keith-turner commented on test/src/main/java/org/apache/accumulo/test/fate/MultipleStoresIT.java:
```diff
@@ -0,0 +1,553 @@
+// TODO 4131 could potentially have separate classes for testing MetaFateStore and UserFateStore
+// similar to how FateTestRunner is used, however that interface doesn't work as nicely here
+// since we are using multiple stores instead of just one. Can do something similar to
+// FateTestRunner here if desired
```
This is interesting.   Wonder if we could pass a factory to test like the following.

```
interface TestFateStoreFactory {
    FateStore create(ZooUtil.LockID lockid);
}
```

Then test could use the factory to create as many stores as needed.  We could run the test passing in an impl that creates MetaFateStore and another that creates UserFateStore.

kevinrr888 commented:
I have compiled and wrote all the follow on issues for this to be merged. This covers all the issues I marked in this PR as `TODO 4131` and all the unresolved review comments. 

_**Here is a synopsis of the follow on issues I will create:**_

* Refactor MultipleStoresIT to function more similarly to other Fate tests like FateIT
* Add FateKey toString() method
* Move MetaFateStore to org.apache.accumulo.core.fate.zookeeper
* Deprecate AbstractFateStore.createDummyLockID(): this includes creating a path in ZK for utilities to get a ZK lock, changing `admin fate fail` and `admin fate delete` commands to get a LockID at this path and no longer require the Manager to be down, and only use a LockID for a store if write operations are expected: the store should fail on write operations if writes are not expected.
* Replace AFS.verifyReserved with a condition on the RESERVATION_COLUMN to verify that it is reserved
* Make WorkFinder and the TransactionRunners critical to the Manager
* Refactor how the RESERVATION_COLUMN works for UserFateStore: create/delete column on reserve/unreserve
* Additional fate test case from https://github.com/apache/accumulo/pull/4524#discussion_r1702301996

_**Here are some questions I still have:**_

1) This was an existing TODO that you had marked in your "distributed FATE" WIP PR:
```
// TODO 4131
// TODO make the max time a function of the number of concurrent callers, as the number of
// concurrent callers increases then increase the max wait time
// TODO could support signaling within this instance for known events
// TODO made the maxWait low so this would be responsive... that may put a lot of load in
// the case there are lots of things waiting...
```
I made the max wait = curr num callers in seconds. Does this address the first TODO? For the other two TODOs, I was not sure what was to be done, so I left them. Can these be safely deleted without creating follow on issue(s)? If follow on issues should be made, perhaps it would be best for you to write these or explain it here because I am not sure what should be done.

2) Re one of your review comments:
"Not a change to make in this PR. Thinking if we push all of the Fate data into a single node it would be nice to use JSON. That would make the data stored in ZK more human readable and it would make it easier to serialize and de-serialize."
Could you elaborate on what you mean by "all of the Fate data into a single node". By "all" do you mean the Repos, TxInfo, Status, Reservation, and FateKey would be in one ZK node? Writing the issue, I want to be sure I'm understanding correctly.

3) Can https://github.com/apache/accumulo/pull/4524#discussion_r1663064467 be resolved? Or should an issue be made?

4) There was a TODO regarding whether the ZooUtil.LockID was the best type for the lock. Should I make a follow on issue regarding this or is ZooUtil.LockID okay?

5) Can this https://github.com/apache/accumulo/pull/4524#discussion_r1663064467 be resolved or should a follow on issue be made?

6) Can these be resolved: https://github.com/apache/accumulo/pull/4524#discussion_r1670934639, https://github.com/apache/accumulo/pull/4524#discussion_r1695958819

7) The DeadReservationCleaner runs every 30 seconds. Is this okay? Should this be longer? Shorter?

Once the above questions are answered, I can complete my full list of new follow-on issues. This will allow me to remove all the `TODO 4131` in this PR and all the comments in this PR can be resolved as all of these will be addressed or have a follow on issue created for them.

keith-turner commented:
I resolved a few of the comments.

> 4. There was a TODO regarding whether the ZooUtil.LockID was the best type for the lock. Should I make a follow on issue regarding this or is ZooUtil.LockID okay?

That type seems like the best exiting type to used, so I do not think an issue is needed.

> The DeadReservationCleaner runs every 30 seconds. Is this okay? Should this be longer? Shorter?

30 seconds seems too short to me.  Would be doing a lot of work and usually never finding anything, should probably be longer maybe 2 or 3 minutes.

> Could you elaborate on what you mean by "all of the Fate data into a single node". By "all" do you mean the Repos, TxInfo, Status, Reservation, and FateKey would be in one ZK node? Writing the issue, I want to be sure I'm understanding correctly.

I was thinking of putting all data that is related to a single fate id into a single zk node.  This will allow atomic updates.  It will be slower (always have to rewrite all repos when adding/removing).  We are only using it for the metadata table.  So thinking it will make things more correct at the expense of speed but that tradeoff is good for metadata related fates.

> Can these be safely deleted without creating follow on issue(s)? If follow on issues should be made, perhaps it would be best for you to write these or explain it here because I am not sure what should be done.

Yeah you can delete those.  I will look into it and open an issue if needed.

keith-turner commented:
@kevinrr888 it seems like the only unresolved conversations are covered by your list of issues to open.  I just left those open for now.

kevinrr888 commented:
The `TODO 4131` comments have now been removed and I believe this PR is ready to be merged :tada:
I was planning on creating the follow-on issues after this is merged in as a batch of new issues.

As suggested by @ctubbsii in our meeting today, I will post a list of all the outstanding changes to be made below. The trivial ones, I will just make a PR for without creating an issue, the more involved follow-ons will be tracked in an issue.

**_Here is the list of changes to be made:_**

_To be tracked by issue:_
- [ ] Refactor MultipleStoresIT to function more similarly to other Fate tests like FateIT
- [ ] Deprecate AbstractFateStore.createDummyLockID(): this includes creating a path in ZK for utilities to get a ZK lock, changing `admin fate fail` and `admin fate delete` commands to get a LockID at this path and no longer require the Manager to be down, and only use a LockID for a store if write operations are expected: the store should fail on write operations if writes are not expected.
- [ ] A single ZK node for all fate data for each fate id
- [ ] Replace AFS.verifyReserved with a condition on the RESERVATION_COLUMN to verify that it is reserved
- [ ] Make WorkFinder and the TransactionRunners critical to the Manager
- [ ] Refactor how the RESERVATION_COLUMN works for UserFateStore: create/delete column on reserve/unreserve

_Will do in one or a few small PRs:_
- [ ] Add a toString() to FateKey
- [ ] Move MetaFateStore to org.apache.accumulo.core.fate.zookeeper
- [ ] Periodic clean up of dead reservations should be increased from every 30 seconds to every few minutes
- [ ] Add additional fate test case suggested in https://github.com/apache/accumulo/pull/4524#discussion_r1702301996

kevinrr888 commented:
Resolved the merge conflicts with `AbstractFateStore` using the new `CountDownTimer`.
Also noticed a sync block that should have been removed so removed that.

`AbstractFateStore` should probably be looked at again before this is merged.

keith-turner reviewed (approved):
Took a look at the changes since my last review and those all look good.

kevinrr888 commented:
@kevinrr888 -
> This was an existing TODO that you had marked in your "distributed FATE" WIP PR:
> // TODO 4131
> // TODO make the max time a function of the number of concurrent callers, as the number of
> // concurrent callers increases then increase the max wait time
> // TODO could support signaling within this instance for known events
> // TODO made the maxWait low so this would be responsive... that may put a lot of load in
> // the case there are lots of things waiting...
> I made the max wait = curr num callers in seconds. Does this address the first TODO? For the other two TODOs, I was not sure what was to be done, so I left them. Can these be safely deleted without creating follow on issue(s)? If follow on issues should be made, perhaps it would be best for you to write these or explain it here because I am not sure what should be done.

@keith-turner -
> Yeah you can delete those. I will look into it and open an issue if needed.

This may still need to be looked into just posting this as a reminder just in case
